In [46]:
import pandas as pd
import numpy as np

pd.options.display.float_format = '{:0f}'.format

In [47]:
df = pd.read_csv('missing_data_practice_lecture.csv')

In [48]:
df.head()

Name       Age       Salary Department
0    Alice 25.000000 50000.000000         HR
1      Bob       NaN 54000.000000         IT
2  Charlie 30.000000          NaN    Finance
3    David 28.000000 62000.000000        NaN
4      Eve 45.000000          NaN  Marketing

In [49]:
df.shape

(27, 4)

In [50]:
df.describe(include='all')

Name        Age           Salary Department
count      27  19.000000        18.000000         21
unique     23        NaN              NaN          5
top     Alice        NaN              NaN    Finance
freq        2        NaN              NaN          6
mean      NaN  53.789474   5632833.388889        NaN
std       NaN  46.100830  23551303.762130        NaN
min       NaN  23.000000     -1000.000000        NaN
25%       NaN  31.000000     52250.000000        NaN
50%       NaN  40.000000     56500.000000        NaN
75%       NaN  45.000000     61000.000000        NaN
max       NaN 210.000000 100000000.000000        NaN

findings:
* negative salary
* really high salary
* missing data - salary, age, department
* ???? - duplicate w/name
* age = 210 - fixed

Missing department - ask Depts about these people
Missing salary - can we get the mean?

In [51]:
df[df['Salary'].isna()]

Name       Age  Salary Department
2   Charlie 30.000000     NaN    Finance
4       Eve 45.000000     NaN  Marketing
6     Grace 35.000000     NaN        NaN
9      Jack 23.000000     NaN         HR
11     Liam       NaN     NaN  Marketing
13     Nina 40.000000     NaN        NaN
15     Paul 33.000000     NaN         IT
17   Rachel 44.000000     NaN  Marketing
20     Nina 40.000000     NaN        NaN

In [52]:
df.dropna(how='all')

Name        Age           Salary Department
0     Alice  25.000000     50000.000000         HR
1       Bob        NaN     54000.000000         IT
2   Charlie  30.000000              NaN    Finance
3     David  28.000000     62000.000000        NaN
4       Eve  45.000000              NaN  Marketing
5     Frank        NaN     57000.000000      Sales
6     Grace  35.000000              NaN        NaN
7      Hank  41.000000     64000.000000    Finance
8       Ivy        NaN     52000.000000         IT
9      Jack  23.000000              NaN         HR
10     Kate 137.000000     56000.000000        NaN
11     Liam        NaN              NaN  Marketing
12      Mia  29.000000     58000.000000      Sales
13     Nina  40.000000              NaN        NaN
14    Oscar        NaN     61000.000000    Finance
15     Paul  33.000000              NaN         IT
16    Quinn        NaN     53000.000000        NaN
17   Rachel  44.000000              NaN  Marketing
18    Steve        NaN     59000.000000      Sales
19     Tina  32.000000    600000.000000         HR
20     Nina  40.000000              NaN        NaN
21    Oscar        NaN     61000.000000    Finance
22    Alice  35.000000     50000.000000         HR
23      Bob  45.000000     55000.000000         IT
24   Johnny 210.000000     -1000.000000      Sales
25      Moe  75.000000         1.000000    Finance
26    Larry  75.000000 100000000.000000    Finance

In [53]:
df.iloc[13]

Name              Nina
Age          40.000000
Salary             NaN
Department         NaN
Name: 13, dtype: object

In [54]:
# all = all columns in subset "and"

df.dropna(subset=['Salary','Department'], how='all')

Name        Age           Salary Department
0     Alice  25.000000     50000.000000         HR
1       Bob        NaN     54000.000000         IT
2   Charlie  30.000000              NaN    Finance
3     David  28.000000     62000.000000        NaN
4       Eve  45.000000              NaN  Marketing
5     Frank        NaN     57000.000000      Sales
7      Hank  41.000000     64000.000000    Finance
8       Ivy        NaN     52000.000000         IT
9      Jack  23.000000              NaN         HR
10     Kate 137.000000     56000.000000        NaN
11     Liam        NaN              NaN  Marketing
12      Mia  29.000000     58000.000000      Sales
14    Oscar        NaN     61000.000000    Finance
15     Paul  33.000000              NaN         IT
16    Quinn        NaN     53000.000000        NaN
17   Rachel  44.000000              NaN  Marketing
18    Steve        NaN     59000.000000      Sales
19     Tina  32.000000    600000.000000         HR
21    Oscar        NaN     61000.000000    Finance
22    Alice  35.000000     50000.000000         HR
23      Bob  45.000000     55000.000000         IT
24   Johnny 210.000000     -1000.000000      Sales
25      Moe  75.000000         1.000000    Finance
26    Larry  75.000000 100000000.000000    Finance

In [55]:
df.loc[df['Department']=='HR','Salary'] = df[df['Department']=='HR'].fillna(50000)

In [56]:
df.loc[df['Department']=='HR','Salary'] = df.loc[df['Department']=='HR', 'Salary'].fillna(50000)

In [57]:
df.loc[df['Department']=='HR','Salary']

0     50000.000000
9     50000.000000
19   600000.000000
22    50000.000000
Name: Salary, dtype: float64

In [58]:
df.loc[df['Name']=='Larry','Salary'] = 100000

In [59]:
df

Name        Age        Salary Department
0     Alice  25.000000  50000.000000         HR
1       Bob        NaN  54000.000000         IT
2   Charlie  30.000000           NaN    Finance
3     David  28.000000  62000.000000        NaN
4       Eve  45.000000           NaN  Marketing
5     Frank        NaN  57000.000000      Sales
6     Grace  35.000000           NaN        NaN
7      Hank  41.000000  64000.000000    Finance
8       Ivy        NaN  52000.000000         IT
9      Jack  23.000000  50000.000000         HR
10     Kate 137.000000  56000.000000        NaN
11     Liam        NaN           NaN  Marketing
12      Mia  29.000000  58000.000000      Sales
13     Nina  40.000000           NaN        NaN
14    Oscar        NaN  61000.000000    Finance
15     Paul  33.000000           NaN         IT
16    Quinn        NaN  53000.000000        NaN
17   Rachel  44.000000           NaN  Marketing
18    Steve        NaN  59000.000000      Sales
19     Tina  32.000000 600000.000000         HR
20     Nina  40.000000           NaN        NaN
21    Oscar        NaN  61000.000000    Finance
22    Alice  35.000000  50000.000000         HR
23      Bob  45.000000  55000.000000         IT
24   Johnny 210.000000  -1000.000000      Sales
25      Moe  75.000000      1.000000    Finance
26    Larry  75.000000 100000.000000    Finance

In [60]:
df[df['Name']=='Johnny']

Name        Age       Salary Department
24  Johnny 210.000000 -1000.000000      Sales

In [61]:
df.iloc[21]

Name                Oscar
Age                   NaN
Salary       61000.000000
Department        Finance
Name: 21, dtype: object

In [62]:
df.duplicated()

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20     True
21     True
22    False
23    False
24    False
25    False
26    False
dtype: bool

In [63]:
df[df['Name']=='Alice']

Name       Age       Salary Department
0   Alice 25.000000 50000.000000         HR
22  Alice 35.000000 50000.000000         HR

In [64]:
df[df.duplicated(subset=['Name','Department'],keep='last')]

Name       Age       Salary Department
0   Alice 25.000000 50000.000000         HR
1     Bob       NaN 54000.000000         IT
13   Nina 40.000000          NaN        NaN
14  Oscar       NaN 61000.000000    Finance

In [65]:
df[df['Name']=='Bob']

Name       Age       Salary Department
1   Bob       NaN 54000.000000         IT
23  Bob 45.000000 55000.000000         IT

In [66]:
df.replace({'IT':'Information Technology'})

Name        Age        Salary              Department
0     Alice  25.000000  50000.000000                      HR
1       Bob        NaN  54000.000000  Information Technology
2   Charlie  30.000000           NaN                 Finance
3     David  28.000000  62000.000000                     NaN
4       Eve  45.000000           NaN               Marketing
5     Frank        NaN  57000.000000                   Sales
6     Grace  35.000000           NaN                     NaN
7      Hank  41.000000  64000.000000                 Finance
8       Ivy        NaN  52000.000000  Information Technology
9      Jack  23.000000  50000.000000                      HR
10     Kate 137.000000  56000.000000                     NaN
11     Liam        NaN           NaN               Marketing
12      Mia  29.000000  58000.000000                   Sales
13     Nina  40.000000           NaN                     NaN
14    Oscar        NaN  61000.000000                 Finance
15     Paul  33.000000           NaN  Information Technology
16    Quinn        NaN  53000.000000                     NaN
17   Rachel  44.000000           NaN               Marketing
18    Steve        NaN  59000.000000                   Sales
19     Tina  32.000000 600000.000000                      HR
20     Nina  40.000000           NaN                     NaN
21    Oscar        NaN  61000.000000                 Finance
22    Alice  35.000000  50000.000000                      HR
23      Bob  45.000000  55000.000000  Information Technology
24   Johnny 210.000000  -1000.000000                   Sales
25      Moe  75.000000      1.000000                 Finance
26    Larry  75.000000 100000.000000                 Finance

In [67]:
df['Name'].replace('M','MMM')

0       Alice
1         Bob
2     Charlie
3       David
4         Eve
5       Frank
6       Grace
7        Hank
8         Ivy
9        Jack
10       Kate
11       Liam
12        Mia
13       Nina
14      Oscar
15       Paul
16      Quinn
17     Rachel
18      Steve
19       Tina
20       Nina
21      Oscar
22      Alice
23        Bob
24     Johnny
25        Moe
26      Larry
Name: Name, dtype: object

In [68]:
df['Name'].str.replace('M','MMM')

0       Alice
1         Bob
2     Charlie
3       David
4         Eve
5       Frank
6       Grace
7        Hank
8         Ivy
9        Jack
10       Kate
11       Liam
12      MMMia
13       Nina
14      Oscar
15       Paul
16      Quinn
17     Rachel
18      Steve
19       Tina
20       Nina
21      Oscar
22      Alice
23        Bob
24     Johnny
25      MMMoe
26      Larry
Name: Name, dtype: object

In [69]:
df['Name'].str.upper()

0       ALICE
1         BOB
2     CHARLIE
3       DAVID
4         EVE
5       FRANK
6       GRACE
7        HANK
8         IVY
9        JACK
10       KATE
11       LIAM
12        MIA
13       NINA
14      OSCAR
15       PAUL
16      QUINN
17     RACHEL
18      STEVE
19       TINA
20       NINA
21      OSCAR
22      ALICE
23        BOB
24     JOHNNY
25        MOE
26      LARRY
Name: Name, dtype: object

In [70]:
df['Name'].replace('Steve', 'steve', regex=True)

0       Alice
1         Bob
2     Charlie
3       David
4         Eve
5       Frank
6       Grace
7        Hank
8         Ivy
9        Jack
10       Kate
11       Liam
12        Mia
13       Nina
14      Oscar
15       Paul
16      Quinn
17     Rachel
18      steve
19       Tina
20       Nina
21      Oscar
22      Alice
23        Bob
24     Johnny
25        Moe
26      Larry
Name: Name, dtype: object

In [71]:
df.rename(columns = str.upper)

NAME        AGE        SALARY DEPARTMENT
0     Alice  25.000000  50000.000000         HR
1       Bob        NaN  54000.000000         IT
2   Charlie  30.000000           NaN    Finance
3     David  28.000000  62000.000000        NaN
4       Eve  45.000000           NaN  Marketing
5     Frank        NaN  57000.000000      Sales
6     Grace  35.000000           NaN        NaN
7      Hank  41.000000  64000.000000    Finance
8       Ivy        NaN  52000.000000         IT
9      Jack  23.000000  50000.000000         HR
10     Kate 137.000000  56000.000000        NaN
11     Liam        NaN           NaN  Marketing
12      Mia  29.000000  58000.000000      Sales
13     Nina  40.000000           NaN        NaN
14    Oscar        NaN  61000.000000    Finance
15     Paul  33.000000           NaN         IT
16    Quinn        NaN  53000.000000        NaN
17   Rachel  44.000000           NaN  Marketing
18    Steve        NaN  59000.000000      Sales
19     Tina  32.000000 600000.000000         HR
20     Nina  40.000000           NaN        NaN
21    Oscar        NaN  61000.000000    Finance
22    Alice  35.000000  50000.000000         HR
23      Bob  45.000000  55000.000000         IT
24   Johnny 210.000000  -1000.000000      Sales
25      Moe  75.000000      1.000000    Finance
26    Larry  75.000000 100000.000000    Finance